Migrate to Quarkus 3 and Java 17

beatngu13 opened the pull request:
This PR migrates the Quarkus Petclinic to Quarkus 3 and Java 17.

I did the Quarkus 3 upgrade with the `quarkus-maven-plugin`. Afterwards, I was facing the following build error:

```text
Caused by: io.quarkus.builder.BuildException: 
Build failure: Build failed due to errors
	[error]: Build step io.quarkus.rest.client.reactive.deployment.RestClientReactiveProcessor#addRestClientBeans threw an exception: java.lang.RuntimeException: The class jakarta.ws.rs.core.Application is not inside the Jandex index
	at org.jboss.resteasy.reactive.common.processor.JandexUtil.isImplementorOf(JandexUtil.java:397)
	at io.quarkus.rest.client.reactive.deployment.RestClientReactiveProcessor.addRestClientBeans(RestClientReactiveProcessor.java:526)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:909)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:282)
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538)
	at java.base/java.lang.Thread.run(Thread.java:833)
	at org.jboss.threads.JBossThread.run(JBossThread.java:501)
```

Which is why I added a `quarkus.index-dependency` entry for `jakarta.ws.rs:jakarta.ws.rs-api`.

For the Java 17 upgrade, I only modified the `pom.xml` so far. There are still these Java 11 references:

1. https://github.com/redhat-developer-demos/quarkus-petclinic/blob/master/devfile.yaml#L17
2. https://github.com/redhat-developer-demos/quarkus-petclinic/blob/master/src/main/docker/Dockerfile.jvm#L26
3. https://github.com/redhat-developer-demos/quarkus-petclinic/blob/master/src/main/docker/Dockerfile.fast-jar#L26

Unfortunately, I'm not quite sure which (Red Hat) images / packages to use for Java 17. Therefore, I marked this PR as a draft.

beatngu13 committed:
Migrate to Quarkus 3

beatngu13 committed:
Migrate to Java 17

geoand commented on src/main/resources/application.properties:
```diff
@@ -20,3 +20,6 @@ quarkus.kubernetes.service-type=load-balancer
 quarkus.http.enable-decompression=true
 quarkus.qute.strict-rendering=true
+
+quarkus.index-dependency.jakarta-ws-rs-api.group-id=jakarta.ws.rs
+quarkus.index-dependency.jakarta-ws-rs-api.artifact-id=jakarta.ws.rs-api
```
I don't really understand why this change. Can you elaborate?

beatngu13 replied:
Me neither 🤷‍♂️ When I did the upgrade:

```text
$ ./mvnw io.quarkus.platform:quarkus-maven-plugin:3.1.3.Final:update -N
```

The build (`./mvnw verify`) was failing afterwards with the error mentioned in the PR description:

```text
Caused by: io.quarkus.builder.BuildException: 
Build failure: Build failed due to errors
	[error]: Build step io.quarkus.rest.client.reactive.deployment.RestClientReactiveProcessor#addRestClientBeans threw an exception: java.lang.RuntimeException: The class jakarta.ws.rs.core.Application is not inside the Jandex index
	at org.jboss.resteasy.reactive.common.processor.JandexUtil.isImplementorOf(JandexUtil.java:397)
	at io.quarkus.rest.client.reactive.deployment.RestClientReactiveProcessor.addRestClientBeans(RestClientReactiveProcessor.java:526)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:909)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:282)
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538)
	at java.base/java.lang.Thread.run(Thread.java:833)
	at org.jboss.threads.JBossThread.run(JBossThread.java:501)
```

I'm not familiar with Jandex, so I tried following [this](https://quarkus.io/guides/cdi-reference#how-to-generate-a-jandex-index) guide from the docs. To my understanding, the `jandex-maven-plugin` doesn't help in this case (tried it anyway, indeed didn't help). But adding this `quarkus.index-dependency` entry fixed the error.

Do you have an explanation for this?

geoand replied:
What happens if you remove the properties?

beatngu13 replied:
As I said, the aforementioned error is thrown.

To illustrate this, I have created a reproducer in a new branch [`quarkus3-java17-no-props`](https://github.com/beatngu13/quarkus-petclinic/commits/quarkus3-java17-no-props) based on this PR. First, I let GitHub Actions build all branches:

* Commit: https://github.com/beatngu13/quarkus-petclinic/commit/626d8f01a060a1a9da6ed14a3c40eab76b766fec
* Successful build: https://github.com/beatngu13/quarkus-petclinic/actions/runs/5462063332/jobs/9940899374

Then, I remove the two properties:

* Commit: https://github.com/beatngu13/quarkus-petclinic/commit/41d6fd013e48acbc497ba3a844b99f6ccbfa686b
* Failed build: https://github.com/beatngu13/quarkus-petclinic/actions/runs/5462092303/jobs/9940966983#step:4:3116

geoand replied:
I will try and have a look tomorrow

geoand replied:
This is actually a bug in Quarkus itself for which I have opened a [fix](https://github.com/quarkusio/quarkus/pull/34570).

Hopefully the fix will land in Quarkus 3.2.2 at which time we can move this PR forward

beatngu13 committed:
Update Quarkus 3.1.3 to 3.2.2

beatngu13 committed:
Remove quarkus.index-dependency entry

Fixed in Quarkus 3.2.1: https://github.com/quarkusio/quarkus/pull/34570

beatngu13 commented:
@geoand updated Quarkus 3.1.3 to 3.2.2, which includes your fix (landed in 3.2.1). Build works fine now. 👍

IMO what's left for this PR to be merge-ready:

>For the Java 17 upgrade, I only modified the `pom.xml` so far. There are still these Java 11 references:
>
>1. https://github.com/redhat-developer-demos/quarkus-petclinic/blob/master/devfile.yaml#L17
>2. https://github.com/redhat-developer-demos/quarkus-petclinic/blob/master/src/main/docker/Dockerfile.jvm#L26
>3. https://github.com/redhat-developer-demos/quarkus-petclinic/blob/master/src/main/docker/Dockerfile.fast-jar#L26
>
>Unfortunately, I'm not quite sure which (Red Hat) images / packages to use for Java 17.

geoand commented:
Nice!

As for the Java 17 image, you can use `java-17-openjdk-headless`

beatngu13 committed:
Use java-17-openjdk-headless for Dockerfiles

beatngu13 committed:
Use udi-rhel8 for devfile

beatngu13 committed:
Update Quarkus 3.2.2 to 3.7.4

beatngu13 commented:
Hi @geoand,

Is there anything I can do, so we can move forward with this PR?

geoand commented:
Sorry, this feel off my radar completely!

Mind updating to 3.8.1 now that we have released it?

beatngu13 committed:
Update Quarkus 3.7.4 to 3.8.1

geoand reviewed (approved):
Thanks!

beatngu13 commented:
Looks like the native image build is failing, will look into that.

beatngu13 committed:
Fix vet_specialties

The DDL of `vet_specialties` is:

```sql
create table vet_specialties (
    specialty_id bigint not null,
    vet_id bigint not null,
    primary key (specialty_id, vet_id)
)
```

The `import.sql` was creating tuples of `(vet_id, specialty_id)`, causing the following exception on startup:

```text
org.postgresql.util.PSQLException: ERROR: insert or update on table "vet_specialties" violates foreign key constraint "fk35uiboyrpfn1bndrr5jorcj0m"
  Detail: Key (specialty_id)=(1004) is not present in table "specialties".
```

beatngu13 committed:
Update Quarkus 3.8.1 to 3.8.2

beatngu13 commented:
The [error message](https://github.com/redhat-developer-demos/quarkus-petclinic/actions/runs/8124893060/job/22206933495#step:4:3482):

```text
 Caused by: com.oracle.graal.pointsto.constraints.UnsupportedFeatureException: Discovered unresolved method during parsing: org.quarkus.samples.petclinic.owner.Pet.getVisits(). This error is reported at image build time because class org.quarkus.samples.petclinic.owner.Pet_ValueResolver is registered for linking at image build time by command line and command line.
```

This similar to https://github.com/quarkusio/quarkus/issues/6396, but that has been resolved in Quarkus 1.2.0 … will further investigate.
